src/NuGet.Services.Metadata.Catalog.Monitoring/Validation/Test/ValidationContext.cs

```diff
@@ -63,34 +71,47 @@ public class ValidationContext
+
+            if (logger == null)
+            {
+                throw new ArgumentNullException(nameof(logger));
+            }
+
             Package = package ?? throw new ArgumentNullException(nameof(package));
             Entries = entries.ToList();
             DeletionAuditEntries = deletionAuditEntries.ToList();
             Client = client ?? throw new ArgumentNullException(nameof(client));
             CancellationToken = token;
 
-            _boolCache = new ConcurrentDictionary<string, Lazy<Task<bool>>>();
-            _indexCache = new ConcurrentDictionary<string, Lazy<Task<PackageRegistrationIndexMetadata>>>();
-            _leafCache = new ConcurrentDictionary<string, Lazy<Task<PackageRegistrationLeafMetadata>>>();
-        }
+            _timestampMetadataResourceV2 = feedToSource[FeedType.HttpV2].GetResource<IPackageTimestampMetadataResource>();
+            _v2Resource = feedToSource[FeedType.HttpV2].GetResource<IPackageRegistrationMetadataResource>();
+            _v3Resource = feedToSource[FeedType.HttpV3].GetResource<IPackageRegistrationMetadataResource>();
 
-        public Task<bool> GetCachedResultAsync(string key, Lazy<Task<bool>> lazyTask)
-        {
-            return _boolCache.GetOrAdd(key, lazyTask).Value;
-        }
+            var commonLogger = logger.AsCommon();
 
-        public Task<PackageRegistrationIndexMetadata> GetCachedResultAsync(
-            string key,
-            Lazy<Task<PackageRegistrationIndexMetadata>> lazyTask)
-        {
-            return _indexCache.GetOrAdd(key, lazyTask).Value;
+            _v2Index = new Lazy<Task<PackageRegistrationIndexMetadata>>(
+                () => _v2Resource.GetIndexAsync(Package, commonLogger, CancellationToken));
+            _v3Index = new Lazy<Task<PackageRegistrationIndexMetadata>>(
+                () => _v3Resource.GetIndexAsync(Package, commonLogger, CancellationToken));
+
+            _v2Leaf = new Lazy<Task<PackageRegistrationLeafMetadata>>(
+                () => _v2Resource.GetLeafAsync(Package, commonLogger, CancellationToken));
+            _v3Leaf = new Lazy<Task<PackageRegistrationLeafMetadata>>(
+                () => _v3Resource.GetLeafAsync(Package, commonLogger, CancellationToken));
         }
 
-        public Task<PackageRegistrationLeafMetadata> GetCachedResultAsync(
-            string key,
-            Lazy<Task<PackageRegistrationLeafMetadata>> lazyTask)
+        public Task<PackageRegistrationIndexMetadata> GetIndexV2Async() => _v2Index.Value;
+        public Task<PackageRegistrationIndexMetadata> GetIndexV3Async() => _v3Index.Value;
+        public Task<PackageRegistrationLeafMetadata> GetLeafV2Async() => _v2Leaf.Value;
+        public Task<PackageRegistrationLeafMetadata> GetLeafV3Async() => _v3Leaf.Value;
+
+        public Task<PackageTimestampMetadata> GetTimestampMetadataAsync()
         {
-            return _leafCache.GetOrAdd(key, lazyTask).Value;
+            return _timestampMetadataResourceV2.GetAsync(this);
```

Comment:
1 - This is specifically the `PackageTimestampMetadata` for V2. The V3 `PackageTimestampMetadata` is constructed from the catalog entries in memory. You should change the naming so that this is clear.

2 - Unlike the other metadata fields, this isn't cached right now. You could easily rewrite `PackageTimestampMetadataResourceV2.GetAsync` to take in `Package` and `Client` directly and then align this with the way `GetIndexV2Async` and the other cached metadata are declared (with `Lazy`) so that it's cached..

Reply:
1.  Done.
2.  Done, except I didn't change the `GetAsync(...)` method signature.  You'd also have to pass `DeletionAuditEntries`, but then why not just pass the `ValidationContext` instance as we already do?